### PR TITLE
Save DQM output per lumisection for UL rereco

### PR DESCRIPTION
**Pull request by schneiml**
#### PR description:

This PR enables per-lumisection saving of *all* DQM output for the UL rereco. This wish came up for machine learning studies targeting per-LS data certification or online monitoring.

In prompt reco, we have per-LS DQM output available due to how Tier0 processes only ~1 LS per job, and the per-job DQMIO output files are available. It would be very nice to have the same granularity of data preserved for the UL rereco, which in addition provides the full run2 dataset processed using the same software.

This is safe to do for most of DQM, since HARVESTING will re-generate per run/job statistics from the per-lumisection snapshots. It is not save for legacy modules that modify MEs at non-standard times or that do not support merging correctly; none of these should exist in RECO/DQM step since the threaded migration, but since there are still a few, I apply the change only to the non-legacy MEs.

#### PR validation:

Validated on a modified `136.85` that crosses a lumi boundary: `runTheMatrix.py -l 136.85 --command '--customise_commands "if process.process == \"reHLT\": process.source.skipEvents=cms.untracked.uint32(3800)\n"' --ibeos -t 8`. Will need some special attention during relval since the DQMIO output file size will increase.

Known problems:
- Various "known unstable" plots change behaviour. These change behaviour on trivial changes such as adding debug output, and might be caused by race conditions.
- The behaviour of per-lumi plots in harvesting changes. This can lead to "double counting" in the existing offline per-lumi harvesting code, and cause subtly incorrect results. Since there is very little such code, I manually adapted the places I could find; however, due to the freedom that harvesting code has, I can not guarantee that these are all cases.

### CODE REVIEW

**Comment by andrius-k** [comment]
please test

**Comment by schneiml** [comment]
One of the things that broke in this PR are the `METRate` plots in multiple places. These seem to be the only 15 differences that RelMon observes. Judging by the code that makes them

https://github.com/cms-sw/cmssw/blob/c1e72b4e1ec7274deb52d505425db826f13f6f33/DQMOffline/JetMET/src/METAnalyzer.cc#L1059-L1074

this is not too big of a loss, given this type of operation in `endRun` was not supported since at least the threaded migration of 2015 (it would need to happen in harvesting), and since the code assumes that a lumisection is 90 seconds I would not trust the computed values anyways.

**Comment by schneiml** [comment]
The more detailed bin-to-bin comparison shows 1000's of changes, waiting for another run to get the output uploaded to DQMGUI. I hope some of the PR tests actually cover more than one lumisection, to make this a meaningful test...

**Comment by schneiml** [comment]
The bin-by-bin comparison noticed that the reference histograms stored by CMSSW turned empty. Though, looking at the names, we don't produce these MEs; the reference data must be from run1, since we have not produced e.g. 'TEC/side1' MEs since 2013 (was renamed to 'TEC/MINUS').

Not sure if these references are still used anywhere (in offline), I'll try to get them back but it might not be worth the effort.

**Comment by davidlange6** [comment]
the reference in the IB system is produced from a recent IB - so if thats the reference you are talking about, then I guess there is still some dqm code using the old convention. Or is there some other reference in your tests?

**Comment by schneiml** [comment]
@davidlange6 , no, I am talking about the reference that is handled inside the DQMStore and saved *into* the legacy root file going to the DQMGUI. I believe this mechanism was not used in Run2, except maybe for a few subsystems in online. Consequently, I have very little of an idea how it works, except for the fact that it exists...

**Comment by schneiml** [comment]
Let's try it a bit less aggressively now. `LSbasedMode` does basically what I want, but needs some care with the DQMIO output. I hope nobody uses `DQMRootOutputModule` in legacy mode, I can't really imagine what the semantcis there would be.

**Comment by schneiml** [comment]
Now would also be a good time to address https://github.com/cms-sw/cmssw/issues/24337 .

As far as I can see, the `postGlobalBeginLumi` can be removed, and the reset should happen in `cloneLumiHistograms`. I'll try this as the next step (assuming the per-lumi saving works).

**Comment by schneiml** [comment]
please test

**Comment by schneiml** [comment]
@Dr15Jones is it legal to have a job with no LuminosityBlock transitions? Several harvesting modules in wf 27434 seem to blow up due to that, they assume there is always an `endLuminosityBlock` before an `endJob`, and do their booking in `endLuminosityBlock`.

Now why this got uncovered by this PR is another question, I guess I confused the `DQMRootSource`...

**Comment by schneiml** [comment]
@Dr15Jones how does this even work?
https://github.com/cms-sw/cmssw/blob/1f7fb8b1c4cb37accf7681591c081103f0e96d09/DQMServices/FwkIO/plugins/DQMRootOutputModule.cc#L462-L463
This seems to read the "base" MEs created by the booking calls, which should be cloned & reset at `endRunProduce`:
https://github.com/cms-sw/cmssw/blob/1f7fb8b1c4cb37accf7681591c081103f0e96d09/DQMServices/Core/src/DQMStore.cc#L1945-L1975

I'm afraid I'll go back to forcing on the `getLumiFlag`, that actually worked without triggering too many time bombs...

**Comment by Dr15Jones** [comment]
@schneiml wrote 
> @Dr15Jones is it legal to have a job with no LuminosityBlock transitions? Several harvesting modules in wf 27434 seem to blow up due to that, they assume there is always an endLuminosityBlock before an endJob, and do their booking in endLuminosityBlock.

It is possible to have no LuminosityBlock transition. This can happen if you process a Run containing no LuminosityBlocks, if you configured the job to only process Runs and not LuminosityBlocks or if you have an exception happen during begin Run processing.

**Comment by Dr15Jones** [comment]
@schneiml wrote
> @Dr15Jones how does this even work?

I'm not certain since I didn't write it. Your best chance of understanding is probably to ask Marco.

https://github.com/cms-sw/cmssw/blame/1f7fb8b1c4cb37accf7681591c081103f0e96d09/DQMServices/FwkIO/plugins/DQMRootOutputModule.cc#L462-L463

**Comment by Dr15Jones** [comment]
Before multi-threading, the DQMStore only held one copy of a histogram. I'm guessing the `m_enableMultiThread ? m_run : 0` was meant to cause the system to go back to that behavior.

**Comment by schneiml** [comment]
I "fixed" the discrepancy in the reference MEs by removing the module that loads them; having this legacy (in edm as well as DQM sense) module in the production sequences seems like a very bad idea in general. I hope nobody actually relies on these Run1 reference histograms read from some database entry.

I also conditioned the "force per lumi flag" to MEs booked in not-legacy mode, given legacy modules like `SiStripDaqInfo` will have a hard time dealing with the per-lumisection resets.

If this passes the tests, I think it should be good to go for the UL rereco.

**Comment by schneiml** [comment]
please test

**Comment by schneiml** [comment]
@Dr15Jones regarding runs without LS: Ok, then multiple harvesting modules (by DT, maybe others as well) are buggy in this situation. I will not fix them here, given this is not really related to this PR.

Regarding the output module: I am interested in the transition ordering there. To me it looks like this code should always save empty histograms, which it clearly does not, so either the `writeRun` happens before `endRunProduce` (which seems strange), or I misunderstand the logic in `DQMStore` (which I checked a number of times now).

The issue there is that even with `enableMultiThread`, the end-of-run snapshots are stored with run ID 0, while the "originals" are reset "for the next lumisection" (not that this makes much sense...).

**Comment by schneiml** [comment]
Let's move that `DQMReferenceHistogramRootFileEventSetupAnalyzer` commit to it's own PR, it should not depend on this change.

**Comment by schneiml** [comment]
please test

**Comment by prebello** [comment]
Hi @schneiml  could you please inform if this change is invisible for rereco requests using the usual dqm sequences? should we change something in cmsdrivers for UL rereco?
which are (if any) the consequences for the event processing in general (time and files size)? Would you expect an increase in the time_event and size_event values?
Thank you

**Comment by schneiml** [comment]
@prebello no change in configuration (cmsDriver etc.) is required. It will affect the file sizes and potentially processing times: the DQMIO output from the RECO step will grow from O(1) to O(n) in the number of lumisections per job (as far as I see at most a few per job, so maybe a 10x increase). This should not be a problem, given the DQMIO output is tiny compared to the event files anyways (before 10MB-100MB per job, after maybe 100MB-1000MB?). The HARVESTING jobs reading this data will consequently also take a bit longer, but I think we never worried about processing time in HARVESTING anyways.

**Comment by schneiml** [comment]
The most recent test result is a false positive: it did not trigger the per-lumisection saving since the workflows run single-threaded in the PR test, using `DQMStore` legacy mode, where I disabled the per-lumi saving... Trying to figure out how to force-on `enableMultiThread` for all DQM (but not HARVESTING) jobs.

**Comment by schneiml** [comment]
Now the tests should be able to do something useful. Though I am afraid they will again not catch all problems, since most wf only handle one lumisection in PR tests. 

Code like this 
https://github.com/cms-sw/cmssw/blob/master/DQMOffline/L1Trigger/src/L1TEGammaOffline.cc#L812-L814
Which is currently sort-of illegal (this needs to happen in harvesting) but works will produce different results with per-lumi saving.

I'll check if I can skip a few 1000 events in wf's like 136.85, so they can cross a lumi boundary with the usual 100 events.

**Comment by schneiml** [comment]
please test

**Comment by schneiml** [comment]
The crashes are due to `ElectronMcFakeValidator` and friends not being able to save output, since their base class, `ElectronDqmAnalyzerBase`, explicitly opts out of lumi transitions. I am afraid we'll need to revert large parts of commit 5dc179c4348b74c2a1f87a6ad141e2635d25babe or get creative with the template arguments to make this works.

**Comment by schneiml** [comment]
@Dr15Jones any hints on how to get back `endLuminosityBlockProduce` calls for *all* of DQM? In this PR, I changed the base for all to `DQMLumisEDProducer`, which includes `edm::EndLuminosityBlockProducer`, but it looks a lot like `one::DQMEDAnalyzer<edm::LuminosityBlockCache<edab::Empty>>` derivatives do not get `endLuminosityBlockProduce` called.

**Comment by Dr15Jones** [comment]
The `DQMEDLumisEDProducer` declares the use of `edm::one::WatchLuminosityBlocks`. The `edm::one::WatchLuminosityBlocks` and the `edm::LuminosityBlockCache` option can not be used together. The use of `edm::one::WatchLuminosityBlocks` prevents the framework from using concurrent luminosity blocks in the job. That is why that option does not work with `edm::LuminosityBlockCache`.

@fabiocos this pull request would seriously compromise CMS' future threading efficiency improvements and ongoing work being done by the framework group. Therefore this pull request is strongly apposed by Core.

**Comment by schneiml** [comment]
> @fabiocos this pull request would seriously compromise CMS' future threading efficiency improvements and ongoing work being done by the framework group. Therefore this pull request is strongly apposed by Core.

@Dr15Jones this is considered as a temporary measure to hit the UL rereco timeline (deadline 2019-04-09). The requested feature was considered quite important by PPD (otherwise we risk running a large amount of additional jobs to make the same information available, that we can get  virtually "for free" out of the ReReco). Unless you have any proposals how to make per-lumisection saving possible without making all of DQM watch the transitions, I think we don't have any other option.

On the longer term (next 2-3 months) we plan to completely replace the `DQMStore` implementation, most likely going back to `edm::stream` as a base of `DQMEDAnalyzer`. I want to massively cut back the complexity and do away with _legacy mode_, _one mode_, _threaded mode_, _concurrent mode_ etc. I hope to have a plan written down and presented by next Tuesday.

We had a number of lengthy discussions on what the options are in the past (some recorded in GitHub issues), and I think I voiced my concern about the _explicit_ `edm::one` approach: We won't be able to get down to 0 DQM modules watching lumi transitions (and thereby enable concurrent lumis) simply because the tendency in DQM is to base _everything_ on lumisections. After all, we are certifying data on lumisection granularity...

**Comment by schneiml** [comment]
Ok, lets try to get the tests to pass and then asses the fallout due to the `LuminosityBlockCache` users. Likely they all need manual changes or another mode switch to give them the old behaviour (as for the legacy modules), but that is more complex than I would have liked for this PR. 

For `ElectronDqmAnalyzerBase` things seem easy enough, the behaviour seems to be not actually used...

**Comment by schneiml** [comment]
please test

**Comment by Dr15Jones** [comment]
> We had a number of lengthy discussions on what the options are in the past (some recorded in GitHub issues), and I think I voiced my concern about the explicit edm::one approach: We won't be able to get down to 0 DQM modules watching lumi transitions (and thereby enable concurrent lumis) simply because the tendency in DQM is to base everything on lumisections. After all, we are certifying data on lumisection granularity...

It is possible to use `one` modules using luminosity transitions and still have concurrent LuminosityBlocks. It just requires the `one` moules to use the LuminosityBlockCache. In that case, the framework still feeds the module just one transition at a time, however it no longer guarantees that all events of one LuminosityBlock are seen before events from another LuminosityBlock are fed to the module.

**Comment by Dr15Jones** [comment]
> On the longer term (next 2-3 months) we plan to completely replace the DQMStore implementation, most likely going back to edm::stream as a base of DQMEDAnalyzer. I want to massively cut back the complexity and do away with legacy mode, one mode, threaded mode, concurrent mode etc. I hope to have a plan written down and presented by next Tuesday.

I think that's a great idea.

**Comment by schneiml** [comment]
Ok, another attempt, this time more selectively restricted to `DQMEDAnalyzer` without any options and with a possibility to opt-out per module. I also added _another_ mode to turn on/off this feature, because, we really only want it for the standard offline DQM.

@Dr15Jones re `LuminosityBlockCache`: that is good to know, but not an option for now, due to how a module can only fill one set of histograms at a time. I plan to release that restriction by adding some smartness to the `MonitorElement`, then we can go to `LuminosityBlockCache`s. Whether that happens explicitly inside the module or hidden in `DQMStore` remains to be decided. I assume the same same thing holds for `edm::stream`, just that `edm::stream` also allows concurrent lumis with "normal" lumi transitions, since different streams can be at different lumis?

**Comment by schneiml** [comment]
please test

**Comment by Dr15Jones** [comment]
> I assume the same same thing holds for edm::stream, just that edm::stream also allows concurrent lumis with "normal" lumi transitions, since different streams can be at different lumis?

Correct.

**Comment by schneiml** [comment]
```
[2019-04-01 11:32:51.855135 +0200][Error  ][XRootDTransport   ] [eoscms.cern.ch:1094 #0.0] Authentication with krb5 failed: Seckrb5: Unable to authenticate credentials;; Too many open files (p=xrootd/eoscms.cern.ch@CERN.CH).
01-Apr-2019 11:32:51 CEST  Fallback request to file root://cms-xrd-global.cern.ch//store/data/Run2016F/ZeroBias/ALCARECO/SiStripCalMinBias-18Apr2017-v1/100000/64A096E7-8F36-E711-A815-38EAA78D8FB0.root
[2019-04-01 11:33:01.102934 +0200][Error  ][AsyncSock         ] [eoscms.cern.ch:1094 #0.0] Socket error encountered: [ERROR] Internal error
----- Begin Fatal Exception 01-Apr-2019 11:33:23 CEST-----------------------
```

This test failure looks unrelated, let's wait for the comparison to see how close we get on the histograms.

**Comment by andrius-k** [comment]
please test with #26283
Failure seems unrelated. Testing with the bin by bin comparison tool.

**Comment by andrius-k** [comment]
please test with #26283
Bin by bin comparison didn't run previously. This is another attempt.

**Comment by schneiml** [comment]
I locally tested a workflow that actually crosses a lumisection boundary:
```
runTheMatrix.py -l 136.85 --command '--customise_commands "if process.process == \"reHLT\": process.source.skipEvents=cms.untracked.uint32(3800)"' --ibeos
```
The bin-to-bin comparison there shows that for some 30000 histograms, the harvesting output does not cover both lumisections. On the other hand, some 100000 histograms are identical before and after the PR...

**Comment by schneiml** [comment]
I am not entirely sure why we had the reset here:
https://github.com/cms-sw/cmssw/blob/a576d2dc75142f1531058dfadc6a214f3ace081f/DQMServices/FwkIO/plugins/DQMRootSource.cc#L630-L635
This is rather surprising semantics to me, I'd expected `setLumiFlag` histos to behave more like online: statistics accumulate until explicitly reset, by default the entire run will be accumulated and saved.

With this reset, the default behavior of `setLumiFlag` histograms is to just save the last lumisection after harvesting... which does not make that much sense.

So I disabled it, and that mostly works (some harvesters do the reset themselves, like they would in online), except now the observable behavior changes and we'll need manual changes on a few subsystem modules (to add the `Reset` calls on MEs that they expect to be reset).

**Comment by schneiml** [comment]
@smuzaffar It seems that this round of tests did not run the bin-to-bin comparison. Shouldn't that work if we test with #26283?

Anyways, we'll need another round of tests with the latest commit, and then I'd really want to see the bin-to-bin comparison.

**Comment by smuzaffar** [comment]
@schneiml , no because we only run it if the python tool is available in release. As it is not yet merged that is why it was not run.

Note that while running comparison we only create cmssw dev area and use tools from the release itself.

**Comment by schneiml** [comment]
A few notes on the plots that change in the latest version of this PR, using the lumisection-transition test from above (http://tinyurl.com/y6xybhzs):

- CSC: I don't really get what changed here, probably entry count, see later.
- DQM: Timing stuff.
- DT, ECAL, L1T, L1TEmu, Muons, RPC, RecoTauV: `DQMEventInfo` seems to do weird things with per lumi saving. Maybe we should opt-out non-TH1 MEs in general, the `DQMRootSource` seems to not know how to handle them anyways.
- EcalBarrel, EcalEndcap: some new plots appeared. Not a problem, I think.
- EcalPreshower: several occupancy maps have lost entries here, probably only last LS is preserved. Not sure how that can happen, maybe there is a `Reset` that was ineffective before somewhere in the harvesting client?
- EGamma: some new plots appeared.
- Egamma / zmumugammaOldValidation / ZToMuMuGamma/nOfflineVtx lost 25 events, thats the first lumi processed.
- HLT/EXO was empty, gained 75 events from the last lumi. Not sure how this happend, but might be a problem. 
- HLT/EventInfo as above.
- HLT/JME some plots gained events, some lost. Weird.
- HLT/SiStrip gained statistics. This might be double counting due to lack of a reset.
- Hcal: several new plots.
- JetMet: values changed. To be investigated.
- SiStrip/EventInfo: as usual
- SiStrip/ControlView: same as in HLT
- Tracking/Eventinfo: as usual
- Tacking/TrackParameters/*/LSanalysis: new plots appear.

**Comment by schneiml** [comment]
What I learned over the last hours of tracing down these differences:
- There are no `Reset` calls on any of the affected MEs.
- All (that I checked) of the problems also happen with all harvesting code disabled; that means no double counting due to lack of resets (in subsystem code at least).

**Comment by andrius-k** [comment]
please test

**Comment by schneiml** [comment]
With the latest change, we are down to a few hundred changing histograms. Some queries and a comparison to a file produced without custom harvesting code running (`HARVESTING:@none`) shows that:
- none of the affected MEs are modified by harvesting code
- The `Hcal/DigiRunHarvesting/Occupancy/depth/depth*` and `Hcal/DigiTask/NumberOfEvents` MEs are the only affected MEs booked in harvesting.

All the others come from the RECO step and are probably only changed in that they now show full run statistics instead of the last lumi.

**Comment by schneiml** [comment]
We have evidence that the Hcal harvesting plots are affected by double counting of lumisections and would need additional `Reset` calls. No other cases found yet.

I am locally testing a more proper implementation of this PR that differentiates between `setLumiFlag` and automatically per-lumi saved MEs. However, this implies a minor change in the DQMIO format, so I am not sure if we want to go with that.

**Comment by schneiml** [comment]
please test

The output seems ok with the independent `CANSAVELUMI` flag, but I have not gotten a local comparison to work. Let's see what the bot thinks...

**Comment by schneiml** [comment]
Went back to the prev. version, and rebased to pre3. The "proper" fix seems to cause trouble in ways that I didn't even know existed, while the problems of the previous version are at least roughly understood...

**Comment by schneiml** [comment]
I can't really track down the unit test failure in the current version (second-to-last test result): It reproduces, but disappears when I add some instrumentation. I'm not really motivated to track this one down, given the IO tests use the legacy mode `DQMStore` with a legacy module and a non-standard reset option for the tests, and I don't really see how any behaviour there should be affected by this PR.

(Ironically, the tests seem to not catch the actual behaviour change in the `DQMRootSource` module...)

**Comment by schneiml** [comment]
The differences in DQM output in this PR can be categorized in a few classes:
- Former per-lumi histograms. For MEs that had the per-lumi flag turned on before this PR, the final saved histogram will have the full data of the run, instead of the values of the last processed lumisection. This is a welcome change, "last processed lumi only" histograms in DQMGUI are dangerously misleading.
- Summaries harvested from these per-lumi plots can be incorrect due to double counting: Since offline harvesting now sees statistics of all previously processed lumisections, earlier lumisections might enter the summary multiple times. To prevent this, a manual `Reset` call is needed in subsystem DQM code. I tried to address this for ECAL in the last PR, but I don't have evidence that it actually works. This is also dangerous, since 0/1 Summaries will not show changed values in any reasonable test data, even if their behaviour might have changed.
- Unexpected per-lumi harvesting: Modules that do harvesting on lumi transitions in offline, but did not have the `setLumiFlag` set, will now suddenly see per-lumi statistics. This will change the output, even though neither the behaviour before nor after this change makes sense.
- Some HLT MET plots show only the first lumisection, even though they had full statisitcs before. This might be a case of the previous problem, but it is not understood.
- In some cases, modules might rely on unsupported behaviour (for example modifying MEs in RECO endJob). These will change behaviour.
- Finally, a large number of changes are minor numerical changes. These are due to reduced rounding errors and floating point saturation problems with per lumi saving enabled: Before, the where many additions of small numbers to large values esp. towards the end of processing, leading to large rounding errors and eventually saturation (to around 17e6 for `TH*F`). This got worse due to the use of one-modules starting last year, which increased the number of events accumulated per histogram. Saving per-lumi histograms and summing up the resulting larger values in harvesting avoids this error.

**Comment by schneiml** [comment]
please test

let's run this version again. We had a very similar one before, but now ECAL and HCAL per-lumi plots should be correctly reset (which shoud result in *empty* output for these plots, instead of the last processed lumi. That is intentional).

**Comment by schneiml** [comment]
Update on the status in this PR: Many of the changes in this PR seem to be caused by DQM code relying on very unstable behaviour, possibly race conditions. These plots change _even when turning on verbose mode_ in `DQMStore`. This includes the ParticleFlow, HLT MET, and many SiStrip plots that showed up in the comparison. Many of them gained additional entries and now show 100 entries for 100 events, which I'd interpret as "broken before, less broken now".

Excluding these, there are another ~150 changes. One fraction of them are the plots with the lumiFlag set, where we intentionally changed the behavior: these are either reset in harvesting (then they will be empty in the output), or they now accumulate and show full run statistics in the output. Both is preferrable to the old default behaviour of showing only one random lumisection. The main consumers here are Tracking, ECAL and HCAL, where I added reset calls after harvesting where I could track down the usages.

Excluding these leaves us with a handfull of mysterious changes: `Egamma/zmumugammaOldValidation/ZToMuMuGamma/nOfflineVtx`, `Tracking/TrackParameters/generalTracks/PUmonitoring/*`, and some probably numerical instabilities. Both groups might also fall under "generally unstable": I see no obvious offending code and many modules are already opted-out of the automatic per-lumi saving.

In general, I'd say this PR is fine to go into a pre-release. The validation was more thorough than what is normally done on DQM infrastructure changes, and we are down to the <1% of weirdly behaving plots. Additional problems may show up during the relvals, since most of the impact of this PR only becomes apparent when many lumisections are processed; but this is in the nature of the task. Overall, this PR will increase the quality of DQM output due to more sane handling of per-lumisection information and reduced numerical errors  when many events are processed in one job, at the risk of affecting a few complex harvesting workflows.

For the future, a larger renovation of the DQM framework in the coming months should take care of the instabilities and inconsistencies that this project uncovered.

**Comment by schneiml** [comment]
please test

let's hope this is the final one!

**Comment by schneiml** [comment]
Turns out there was a bug in the per-lumi saving implementation since the `edm::one` migration from last year, that effectively lead to an empty per-lumi histogram being saved. This is most visible in 1325.7 in the current set of results (first lumi lost ~everywhere), and might also explain some of the other remaining discrepancies. 1a263cf should fix that.

**Comment by schneiml** [comment]
please test

I hope this last fix did not break anything else. Looking forward to even less changes in the next round.

**Comment by andrius-k** [comment]
+1

**Comment by schneiml** [comment]
Will turn the feature off by default once the tests are done, then this one can go in.

**Comment by schneiml** [comment]
@prebello Since the per-lumi saving is now turned off by default, we will need this to be turned on explicitly for a few RelVals and (if all goes well) also rereco jobs. We should get in touch about how to proceed once this PR is merged.

**Comment by schneiml** [comment]
please test

No, with per-lumi saving turned off, I still expect a similar number of output changes: the fallout is mostly not due to the per-lumi saving itself, but due to changed behaviour in harvesting. These changes are unconditional, and as explained before, fix some real problems but also cause some risks.

**Comment by andrius-k** [comment]
+1

**Comment by fabiocos** [comment]
please test

**Comment by schneiml** [comment]
@fabiocos After input I got from Dirk Hufnagel about the DQMIO file merging, this will need another minor fix to make the merge jobs behave correctly. Commit will come in ~minutes.

**Comment by andrius-k** [comment]
please abort

**Comment by andrius-k** [comment]
please test

**Comment by andrius-k** [comment]
+1

File: Configuration/Applications/python/ConfigBuilder.py [hunk 1]
```diff
@@ -1935,6 +1935,8 @@ def expandMapping(self,seqList,mapping,index=None):
     def prepare_DQM(self, sequence = 'DQMOffline'):
         # this one needs replacement
 
+        # any 'DQM' job should use DQMStore in non-legacy mode (but not HARVESTING)
+        self.loadAndRemember("DQMServices/Core/DQMStoreNonLegacy_cff")
```
Comment (fabiocos) [inline Configuration/Applications/python/ConfigBuilder.py]: @schneiml this essentially sets always ```DQMStore.enableMultiThread = True```, and it should not change behaviour for multi-threaded workflows in my understanding, and not prevent HARVESTING to run anyway, right?

File: Configuration/DataProcessing/python/Merge.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def mergeProcess(*inputFiles, **options):
     #//
     if newDQMIO:
         process.source = Source("DQMRootSource")
-        process.add_(Service("DQMStore"))
+        process.add_(Service("DQMStore", forceResetOnBeginLumi = CfgTypes.untracked.bool(True)))
```
Comment (fabiocos) [inline Configuration/DataProcessing/python/Merge.py]: @schneiml if the feature needs to be off by default, should not this flag be turned to False?

**Comment by fabiocos** [comment]
@schneiml could you please address my two questions? I would like to move forward with this PR before pre4 if possible, I just need a clarification

File: Configuration/Applications/python/ConfigBuilder.py (hunk 1 above)
Comment (schneiml) [reply Configuration/Applications/python/ConfigBuilder.py]: This is completely independent of HARVESTING, which should be configured somewhere else.

And yes, all serious workflows run with `enableMultiThread = True` anyways, so it makes sense to turn it on even on single-threaded workflows to get consistent behavior.

File: Configuration/DataProcessing/python/Merge.py (hunk 2 above)
Comment (schneiml) [reply Configuration/DataProcessing/python/Merge.py]: No, the changes in HARVESTING behavior are spread over many places and unconditional. This fixes a number of problems with the existing per-lumi plots (at the risk of causing some misbehavior in said plots, as explained in the PR message). 

Conditional is only the change in the RECO step, which turns on per-lumi saving on all plots (that allow it).

**Comment by fabiocos** [comment]
+operations

**Comment by fabiocos** [comment]
+1

**Pull request merged**